### PR TITLE
Observer pattern event implementation

### PR DESCRIPTION
**Pull request by KaseyCantu**
Publish and Subscribe to RequestSentEvent and ResponseReceivedEvent.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -89,6 +89,29 @@ poetry install
 poetry shell
  ```
 
+### Pre-Commit Hooks
+We are using [Pre-Commit](https://pre-commit.com/) to enforce formatting, lint rules, and code analysis so that
```
Comment (nate-williams-auctane) [inline README.md:93]: This is great documentation, very nice. The local install instructions were very much appreciated.

File: shipengine_sdk/http_client/client.py [hunk 2]
```diff
@@ -13,12 +14,48 @@
...
 
 
+def base_url(config) -> str:
+    return config.base_uri if os.getenv("CLIENT_BASE_URI") is None else os.getenv("CLIENT_BASE_URI")
+
+
+def check_for_errors(status_code: int, response_body: Dict[str, Any], config) -> None:
+    is_response_404(status_code=status_code, response_body=response_body, config=config)
+    is_response_429(status_code=status_code, response_body=response_body, config=config)
+    is_response_500(status_code=status_code, response_body=response_body)
```
Comment (nate-williams-auctane) [inline shipengine_sdk/http_client/client.py]: I might follow up sometime on this out of curiosity, but I think everything looks great overall and read and flowed nicely.

**Review by nate-williams-auctane — APPROVED** [review]
I don't have a lot of background/experience on the Observer Pattern in Python, and have been foreign to python for a good while, but I think this structure made sense to me. I'm interested to see how the tests flush out! Looks good to me.

File: shipengine_sdk/http_client/client.py (hunk 2 above)
Comment (KaseyCantu) [reply shipengine_sdk/http_client/client.py]: Yes, I am happy to sync on this anytime! The `is_response_123` methods (along with a few other methods) were making the client class body enormous so I tried to move things out into functions where I would help that cause. I would ultimately like to have one function `check_response_for_error()` that has all the `is_response_123` logic int

Comment (KaseyCantu) [reply shipengine_sdk/http_client/client.py]: Actually, I might make that change before closing/merging this PR.

Comment (KaseyCantu) [reply shipengine_sdk/http_client/client.py]: Change made! Thanks Nate.

Comment (KaseyCantu) [reply shipengine_sdk/http_client/client.py]: Change made! Thanks Nate.

**Pull request merged**